lib/SILOptimizer/Analysis/ColdBlockInfo.cpp

```diff
@@ -221,31 +238,99 @@ bool ColdBlockInfo::inferFromEdgeProfile(SILBasicBlock *BB) {
   SmallVector<ProfileCounter, 2> succCount;
 
   // Current analysis only accurately handles blocks with 2 successors,
-  // especially since we only have two temperatures.
+  // especially since we only have two temperatures (cold/warm).
+  // TODO: With propagation algorithm, this limitation can be removed to support
+  // multi-way branches (switch statements) and arbitrary successor counts.
   if (BB->getNumSuccessors() != 2)
     return false;
 
-  // Check the successor edges for profile data.
+  // First pass: collect all counters and check for evidence of profiling
+  // Note: std::optional<ProfileCounter> is used to distinguish between:
+  //   - std::nullopt: missing profile data (profiler didn't record this edge)
+  //   - ProfileCounter(0): definitely zero executions (profiler recorded 0)
+  // This distinction is crucial for both the current optimistic strategy and
+  // future propagation algorithms, which need to know where data is missing.
+  bool hasAnyNonZeroCount = false;
+  bool hasAnyMissingData = false;
+  SmallVector<std::optional<ProfileCounter>, 2> counters;
```

Comment:
`std::optional<ProfileCounter>` is redundant here. ProfileCounter already has a representation of "missing"